### ISSUE TITLE
Filtered item visible in List when model has been changed.

### ISSUE THREAD

**Issue by peuter** [qooxdoo/qooxdoo#196]
See Playground example:
http://tinyurl.com/hftbw23

Press the shift button to see the problem

A Listmodel with 5 Items where the item on index 1 is filtered. When you shift the model array the filtered Item is shown in list.

**Comment by level420** [comment]
I think as of http://demo.qooxdoo.org/current/apiviewer/#qx.data.controller.List you have to update the controller after manipulating the model:

http://tinyurl.com/hghqbac

**Comment by peuter** [comment]
Shouldn't the controller update itself after the model has changed?

And 2nd. if you check the console in the playground you can see that the filtered model item gets applied to the list item. This problem also exists when you use virtual lists btw. Your workaround just hides the real problem IMHO.

**Comment by level420** [comment]
You're right. This should happen automatically.It may be, that the fix in https://github.com/qooxdoo/qooxdoo/commit/6570b3d1df776548d9d5af4f82ee2d1229df59fc addresses that issue.

Which version of qooxdoo do you use? Playground is still on qooxdoo 5.0.1 where the fix is not in yet.

**Comment by peuter** [comment]
I started a new project with the current master-branch to investigate the problem (without any success yet). so the problem also exists in the current master branch.

**Comment by level420** [comment]
@johnspackman: This is your fix: https://github.com/qooxdoo/qooxdoo/commit/6570b3d1df776548d9d5af4f82ee2d1229df59fc

Could you please have a look at this?
Thanks!

**Comment by level420** [comment]
OK then let's see: 

If the model.shift() is called in https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/Array.js#L301 the change event should be fired. 

Because the list controller registers an event listner for the model change event here https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/controller/List.js#L328 the controllers method __changeModel should be fired: 

https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/controller/List.js#L404 which calls the list controllers update method at https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/controller/List.js#L412

**Comment by johnspackman** [comment]
i can take a look at this if you want, i've got some hours put aside for Qooxdoo stuff this afternoon?

**Comment by oetiker** [comment]
you write 'should' ? does the event not get fired?

**Comment by level420** [comment]
Sorry, I meant it is fired.

**Comment by level420** [comment]
@johnspackman yes please!

**Comment by peuter** [comment]
I think a narrowed it down a litte bit. The problem is that the Direct dispatcher caches the listeners vefore traversing them (here: https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/event/dispatch/Direct.js#L117). In the playground example the listeners[0].handler.class(context, event) removes some the the later listeners.

If you renre the local listener list by adding: 
`listeners = this._manager.getListeners(target, type, false);` 
 after this line: https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/event/dispatch/Direct.js#L134

The problem is solved. Of course this is just a quick hack, but should help to find a proper solution.

**Comment by peuter** [comment]
I just pushed a possible fix as a starting point for further investigation. Please let me know what you think abount this.

**Comment by johnspackman** [comment]
i'm making some progress with this but it does not appear to be a problem in qx.data.controller.List at all - it's in qx.data.SingleValueBinding.__chainListener, which leaks listeners when dealing with arrays.

Eg if you have an class with model property and a value property, this code...:

```javascript
this.bind("model[0]", this, "value");
```

will leak a binding.  List controller uses exactly this form, but also requires that it is able to unbind all listeners and reapply them, but the leaked binding is overwriting properties; in @peuter's example it is the leaked binding that triggers the "Filtered model item applied"

**Comment by level420** [comment]
@johnspackman What do you mean by "leak a binding"?

**Comment by level420** [comment]
If we decouple the call of the update method within the __changeModel listener within qx.data.controller.List by replacing this line https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/controller/List.js#L412
with
```javascript
      qx.event.Timer.once(this.update, this, 0);
```
the bug seems to be solved.

**Comment by johnspackman** [comment]
Sorry, that wasn't a very complete example!

This code:

```javascript
var id = this.bind("model[0].property", this, "value");
this.removeBinding(id);
```

Should be the same as if the code had never run, because removeBinding() undoes all addListener()'s done by bind(); for the example above, addListener is actually called several times, at least once for each segment in the path and the "id" is an object that contains (among other things) a list of these listener IDs.

However, the problem is that when an array element is bound, there are two listener IDs required, and the second one overwrites the first one.  This means that when you call removeBinding() one listener is still in place and working.

AFAICT this is the underlying cause of the problem with the List controller, because it means that change events fire that should have been removed previously.

My guess is that using a timer as in your example changes the ordering slightly and works around the problem.  The test case I have now doesn't use any of the List controller example, simply binds the model and has an assert in SingleValueBinding to trap addListener IDs being lost.

**Comment by level420** [comment]
Thank you for the explanation John.

**Comment by peuter** [comment]
@johnspackman Could you provide a complete example of yout test case. I tried to repwoduce it with
```
      var model = new listtest.Array();
      for (var i=0; i<3; i++) {
        model.push(new listtest.Dummy().set({name : "Dummy "+i}));
      }
      this.setModel(model);
      var id = this.bind("model[0].name", this, "value");
      this.addListener("changeValue", function(ev) {
        console.log("value changed: "+ev.getData());
      }, this);
      console.log("removing %O",model.shift());
      console.log("removing Binding %s", this.removeBinding(id));
      console.log("removing %O",model.shift());
```
But I don't see the last model change fires any events. 

During debugging I initially also thought that the binding is created twice and overwrites the first ID, leading to something like a "ghost" binding. But AFAIR the SingleValueBinding.__chainListener removes the old binding correctly before creating the new one here https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/SingleValueBinding.js#L305-L316
So I took a look into the event dispatcher in the moment when the "problematic" event is handled and saw that the binding has already been removed in that moment, but as the listeners have been cached before traversing throught the list, the removal isn't recognized there. So I came up with the hack above to see if I was on the right track.

Long story short: I just want to know if I got something wrong here, so your example might help me to find that out.

**Comment by johnspackman** [comment]
It's a bit tricky to that at the moment because I've got loads of other little debugging hacks in there at the moment, but my test was to modify these lines:

https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/SingleValueBinding.js#L375
https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/SingleValueBinding.js#L390
https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/SingleValueBinding.js#L412

to add:
```javascript
qx.core.Assert.assertFalse(!!context.listenerIds[j]);
```

and also:
https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/data/SingleValueBinding.js#L1176

to add:
```javascript
delete id.listenerIds[i];
```

The assert fails because the listenerId is being overwritten, so I assumed that the listenerId is not being cleaned up as a side effect of something else.

However, while I think that's a bug in SingleValueBinding and I have a fix for it, I no longer think that it's the cause of this issue because the problem still occurs :(

**Comment by johnspackman** [comment]
PS - I don't think you've got anything wrong, your use case looks perfectly fine to me, IMHO it's a bug in controller.List or SingleValueBinding

**Comment by peuter** [comment]
Ok thanks the explanation, btw. the problem also occurs when you use virtual Lists so either its a bug in both controllers or in the SingleValueBinding.

**Comment by johnspackman** [comment]
OK I think I've got it - 

The issue is that when the changeModel event starts firing, it gets an array of event handlers that need to be fired and then fires them one by one; when controller.List catches the "changeModel" event (__changeModel) it calls List.update() which removes and rebinds all of the items.  This is intended to make the bindings correctly reflect which model item relates to which ListItem.

However, the list of event handlers to call has already been determined and the (supposedly) unbound event handler is already in that array, and when it fires it has the wrong model because the array has shifted down by one.  IE the handler which controller.List has just unbound (and which is in the queue to be fired) referred to "model[1]" but it now needs to refer to "model[0]"; when the old handler fires, this makes it set the "filtered" Dummy instance to the Item.  

@level420 Dietrich your solution with setTimeout works because it allows the list of event handlers to complete before the unbinding/rebinding happens, although actually this means that the list is set to a wrong value and then corrected; the log still shows "Filtered model item applied" 

The issue is mainly caused because List binds based on explicit array indexes (eg the bind path it gives to SVB is "model[0].name", and the order of events triggering cannot be modified soon enough.

So, as to a solution - IMHO binding via explicit indexes in strings is not ideal and I'm not convinced that there's an easy fix; my instinctive reaction is to remove index based binding all together in favour of binding to properties of the array item (ie bind model.getItem(0) to "propertyName", rather than bind this to "model[0].propertyName") and then watch the model for "change" events for add & remove.  IMHO this is also more efficient because there is a surprisingly large number of addListener/removeListener calls going on.

However the down side would be that this would have a breaking change in the API for qx.data.controller.IControllerDelegate#bindItem, which expects to have an index.

**Comment by oetiker** [comment]
for breaking changes we just need a new method name ... and if unfixable deprecate the old bindItem ... 

even better would be to deprecated AND fix bindItem with a hint on using the new method.

**Comment by level420** [comment]
Maybe one step back again: why does the playground example work, if we do not do any filtering?

http://tinyurl.com/jacasjr

**Comment by johnspackman** [comment]
that's a good question, i'll try and find an answer for you in a shortly.

In the mean time, I've figured out why it is important to List to bind based on model indexes - it's because it supports binding to arrays of scalars, and not just qx.core.Object derived classes.

This is quite a limitation because it completely changes the strategy for binding and introduces complexity; I've rewritten the List controller as List2 but only noticed this when porting the unit test from List, where it turns out that almost all of the tests are based on arrays of scalars.

**Comment by johnspackman** [comment]
Here's where i am on this: https://github.com/johnspackman/qooxdoo/commit/72dd5a8e4d0853f0eac0743fc94f8b873d1e3d3e

Having decided that it's the binding based on model indexes that is causing a problem, this new class (qx.data.controller.List2) binds properties of model items to target items (ie qx.ui.form.ListItem derived objects).  

The changes for binding like this are extensive and I couldn't see a way of making this backwards compatible, so instead opted for a separate class based on the old qx.data.controller.List code so it is a drop in replacement *provided* you don't use delegates.

I've opted to put the array management into a separate class (qx.data.controller.ArrayMirror) because i think it's quite useful for controllers, and the abstraction is quite natural also.  Similarly for filtering, the qx.data.controller.ArrayFilter classes provides filtering and sorting for any qx.data.Array instance - this is actually one of my own classes that I've used for a while and IMHO it's a useful addition in its own right

List2 does not support arrays of scalars though, as mentioned above and I'm debating where to go with this - EG to keep two implementations with different characteristics isn't exactly ideal.

**Comment by oetiker** [comment]
wow you are productive ... I am impressed ... how about calling it QxObjectList or something reminiscent of the non scalar restriction ? List2 seems a bit non descriptive :-)

**Comment by derrell** [comment]
However, to make it more easily found when someone is looking for a list,
how about ListObject or even ListOfObjects (although the latter is a bit
more verbose than is qx-typical). That way, they'll see both of them and
say, "Huh... I wonder what the difference is..." which will, of course, be
well-documented in the class. :-)

Derrell

**Comment by johnspackman** [comment]
thanks folks, I've renamed it (List2 was supposed to be temporary, honest ;) ) and went for ListOfObjects.  I'll submit as a PR shortly.

I've been toying with another wrapper/controller class that maps an array of scalars onto a qx.data.Array of proper objects; this seemed a bit over-engineered at first, but OTOH would allow controller.List to be fully deprecated in favour of a ListOfObjects and ListOfScalars, and removes the need for writing a special scalar-compatible equivalent for other controllers (in fact ListOfScalars would only be necessary so that there is a drop in replacement for List, but it can't replace List because the caller has to decide which variant they need to use)

**Comment by derrell** [comment]
We should really think long and hard about this. Is the problem it's
solving severe enough to warrant this fairly drastic action?

Derrell

**Comment by oetiker** [comment]
just guessing ... without having looked at the code ... how about making .List a wrapper which does some introspection on the arguments it is getting and decides on which controller to use ?

**Comment by derrell** [comment]
+1 !!!

**Comment by johnspackman** [comment]
It could be a wrapper, except that I think it will be impossible to make the delegate work properly because that has binding by array index baked into the interface.  Having said that, there's nothing to bind to unless your array contains objects anyway.

If we can't fix List without changing the API at least a little bit, we could just plan to leave List in place (and with unchanged API) long term instead of deprecating with the plan that it will be removed in the version after next.

I'm quite happy with that TBH - List is pretty mature and stable (a quick look it's history and it only been doc changes for years), and although it was a pain and time consuming to figure out what was going on, we now know.

**Comment by oetiker** [comment]
could there be a way for List to detect that something fishy is happening and at least complain about it ?

**Comment by johnspackman** [comment]
fixed - I've added a one-time warning

**Comment by oetiker** [comment]
so @level420  and @peuter please check if 6964166 helps with your bug

**Comment by level420** [comment]
@oetiker I'm sorry, but I was just driving the discussion and was tying to help without currently having an issue myself, so I hand this ovet to @peuter.

**Comment by peuter** [comment]
I will check that on Monday.

**Comment by peuter** [comment]
Ok the only part of the PR that is relevant for the reported issue IMHO is the qx.event.Timer.once(this.update, this, 0); in qx.data.controller.List and that indeed fixes the problem in the playground example. But as I already mentioned twice, we have the same problem in virtual lists. If you change the playground example to virtual lists, you can still see the problem:
http://tinyurl.com/hdwrmsz

So the PR is only fixing one part of the problem.

**Comment by johnspackman** [comment]
The addition of qx.event.Timer.once does not fix the problem, but it does change the bug so that the net effect is one where the UI will display the correct result; what happens is that the model of the "Item" class will be changed to a filtered out item and then reset to the correct, unfiltered item.  

AFAICT this bug is not going to be fixable in List without modifying the API - specifically, the IControllerDelegate interface would have to change because it works on binding to an array index.  

So you can continue to use the List class with this fix, and in most situations it will be perfectly fine; a warning is issued if there is a possible problem you would need to look at.

I have not looked at Virtual Lists (yet), but I'd suggest getting the first pull request integrated first and looking at virtual lists as a second phase

**Comment by derrell** [comment]
I feel fairly strongly that adding a new class here is a mistake if we have
a viable workaround for the moment. qooxdoo is already difficult enough to
learn that adding more classes, similar to existing ones, seems the wrong
thing to do. I propose that we add the workaround with the timer, which I
understand just masks the bug (but does so in a way that is not generally
user-visible), and *not* add the new class. In the next major release, we
can change the API, and user code will have to change to adapt to the new
API. This is especially true if we're going to have a similar potential
added class for Virtual Lists. I don't like workarounds, but I dislike even
more the thought of further complicating qooxdoo's learning curve.

Derrell

**Comment by peuter** [comment]
I'm also not happy with this workaround as it only hides the problem. I see some usefull things on @johnspackman pull request, but they are not related to this issue. I would suggest to merge the PR without the workaround in the list controller and find a proper solution for the issue.

IMHO the underlying problem here is, that it is possible that the listeners can change during the dispatching process (e.g. because one event handler removes other listeners). But they still get dispatched, because the "listener-list" has been cached before processing. A possible solution might be to introduce a blacklist where the event manager adds the listener ids after they got removed. The dispatchers just have to check if the currently processed listener is blacklisted and do not dispatch it if that is the case. I can try to implement a solution, but would like to discuss it first. Maybe someone has a better idea.

**Comment by johnspackman** [comment]
If there's a solution which works by changing the binding mechanism then I completely agree that would be preferable because it is [supposed to be] legitimate to bind to array indexes, and it's an unexpected side effect of binding that causes the problem.

Also in favour is that it would support binding of arrays of scalars, which cannot be supported with the approach taken by ListOfObects.

Related to this is that List currently has to remove and re-add the listener for model.change (see https://github.com/johnspackman/qooxdoo/blob/issue-196/framework/source/class/qx/data/controller/List.js#L484-492) to keep it at the end of the chain - there's no specific use case for this type of behaviour, but it's clearly a work around for a similar kind of problem.

**Comment by peuter** [comment]
So this is my suggestion to fix the issue. Tested with list.List and form.List. Let my know what you think.

@johnspackman I have seen the __moveChangeListenerAtTheEnd workaround but I don't know if my changes would make that one obsolete, as I don't know which problem it solves. I did a short test by removing it and it didn' t break anything. As it is quite old code, I would only delete it if we have an issue to test against and see if it is obsolete or not.
